src/core/raster/qgsrasterattributetable.cpp

```diff
@@ -0,0 +1,411 @@
+{
+  return insertField( name, usage, type, mFields.count() );
+}
+
+bool QgsRasterAttributeTable::appendField( const Field &field )
+{
+  return insertField( field, mFields.count() );
+}
+
+bool QgsRasterAttributeTable::removeField( const QString &name )
+{
+  const auto toRemove { std::find_if( mFields.begin(), mFields.end(), [ &name ]( Field & f ) -> bool {
+      return f.name == name;
+    } )};
+
+  if ( toRemove != mFields.end() )
+  {
+    const long long idx { std::distance( mFields.begin(), toRemove ) };
+    mFields.erase( toRemove, mFields.end() );
+    for ( auto it = mData.begin(); it != mData.end(); ++it )
+    {
+      it->removeAt( idx );
+    }
+    setIsDirty( true );
+    return true;
+  }
+
+  return false;
+}
+
+bool QgsRasterAttributeTable::insertRow( const QVariantList data, int position )
+{
+  if ( position < 0 )
+  {
+    return false;
+  }
+
+  QVariantList dataValid;
+
+  if ( dataValid.size() > mFields.size() )
+  {
+    for ( int colIdx = 0; colIdx < mFields.size(); colIdx++ )
+    {
+      dataValid.append( data[ colIdx ] );
+    }
+  }
+  else
+  {
+    dataValid = data;
+  }
+  mData.insert( position, dataValid );
+  setIsDirty( true );
+  return true;
+}
+
+bool QgsRasterAttributeTable::appendRow( const QVariantList data )
+{
+  return insertRow( data, mData.count() );
+}
+
+bool QgsRasterAttributeTable::writeToFile( const QString &path, QString *errorMessage )
+{
+  QgsVectorFileWriter::SaveVectorOptions options;
+  options.actionOnExistingFile =  QgsVectorFileWriter::ActionOnExistingFile::CreateOrOverwriteFile;
+  options.driverName = QStringLiteral( "ESRI Shapefile" );
+  options.fileEncoding = QStringLiteral( "UTF-8" );
+  std::unique_ptr<QgsVectorFileWriter> writer;
+
+  // Strip .dbf from path because OGR adds it back
+  QString cleanedPath { path };
+  if ( path.endsWith( QStringLiteral( ".dbf" ), Qt::CaseSensitivity::CaseInsensitive ) )
+  {
+    cleanedPath.chop( 4 );
+  }
+
+  writer.reset( QgsVectorFileWriter::create( cleanedPath, qgisFields(), QgsWkbTypes::Type::NoGeometry, QgsCoordinateReferenceSystem(), QgsCoordinateTransformContext(), options ) );
+
+  const QgsVectorFileWriter::WriterError error { writer->hasError() };
+  if ( error != QgsVectorFileWriter::WriterError::NoError )
+  {
+    if ( errorMessage )
+    {
+      *errorMessage = QObject::tr( "Error creating RAT table: %1." ).arg( writer->errorMessage() );
+    }
+    return false;
+  }
+  auto features { qgisFeatures() };
+  bool result { writer->addFeatures( features ) };
+
+  if ( ! result )
+  {
+    if ( errorMessage )
+    {
+      *errorMessage = QObject::tr( "Error creating RAT table: could not add rows." );
+    }
+    return false;
+  }
+
+  result = writer->flushBuffer();
+
+  if ( result )
+  {
+    setIsDirty( false );
+  }
+
+  return result;
+}
+
+bool QgsRasterAttributeTable::readFromFile( const QString &path, QString *errorMessage )
+{
+  QgsVectorLayer rat { path, QStringLiteral( "rat_temp" ), QStringLiteral( "ogr" ) };
```

Comment:
Could possibly save some overhead here by just creating a QgsGdalProvider instead -- we don't need any of the vector layer methods here